src/lib/eap/base.c

```diff
@@ -386,80 +388,81 @@ void eap_add_reply(request_t *request, fr_dict_attr_t const *da, uint8_t const *
+unlang_action_t eap_virtual_server(request_t *request, eap_session_t *eap_session, char const *virtual_server)
 {
-#if 1
-	return RLM_MODULE_FAIL;
-#else
-	eap_session_t	*eap_session_inner;
-	rlm_rcode_t	rcode;
-	fr_pair_t	*vp;
-	CONF_SECTION	*server_cs;
+	virtual_server_t const	*vs;
+	CONF_SECTION		*server_cs;
+	fr_pair_t		*vp;
 
-	vp = fr_pair_find_by_da(&request->control_pairs, NULL, attr_virtual_server);
-	server_cs = vp ? virtual_server_find(vp->vp_strvalue) : virtual_server_find(virtual_server);
+	fr_assert(request->parent);
 
-	if (server_cs) {
-		RDEBUG2("Running request through virtual server \"%s\"", cf_section_name2(unlang_call_current(request)));
-	} else {
-		RDEBUG2("Running request in virtual server");
-	}
+	vs = virtual_server_find(virtual_server);
+	server_cs = virtual_server_cs(vs);
 
-	/*
-	 *	Add a previously recorded inner eap_session_t back
-	 *	to the request.  This in theory allows infinite
-	 *	nesting, but this is probably limited somewhere.
-	 */
-	if (eap_session->child) {
-		RDEBUG4("Adding eap_session_t %p to child request", eap_session->child);
-		request_data_talloc_add(request, NULL, REQUEST_DATA_EAP_SESSION,
-					eap_session_t, eap_session->child, false, false, false);
+	if (!server_cs) {
+		RERROR("Virtual server \"%s\" missing", virtual_server);
+		return UNLANG_ACTION_FAIL;
 	}
 
-	rad_virtual_server(&rcode, request);
-	eap_session_inner = request_data_get(request, NULL, REQUEST_DATA_EAP_SESSION);
-	if (eap_session_inner) {
-		/*
-		 *	We assume if the inner eap session has changed
-		 *	then the old one has been freed.
-		 */
-		if (!eap_session->child || (eap_session->child != eap_session_inner)) {
-			RDEBUG4("Binding lifetime of child eap_session %p to parent eap_session %p",
-				eap_session_inner, eap_session);
-			talloc_link_ctx(eap_session, eap_session_inner);
-			eap_session->child = eap_session_inner;
-		} else {
-			RDEBUG4("Got eap_session_t %p back unmolested", eap_session->child);
-		}
+	RDEBUG2("Running request through virtual server \"%s\"", virtual_server);
+
 	/*
-	 *	Assume the inner server freed the
-	 *	eap_session_t and remove our reference to it.
-	 *
-	 *	If it didn't actually free the child (due to error)
-	 *	the call to talloc_link_ctx (above) ensures it will
-	 *	be freed when the parent is.
+	 *	Create a State attribute to retrieve the session state
+	 *	stored on the previous round.
 	 */
-	} else if (eap_session->child) {
-		RDEBUG4("Inner server freed eap_session %p", eap_session->child);
-		eap_session->child = NULL;
+	if (eap_session->child_state) {
+		fr_pair_append_by_da(request->request_ctx, &vp, &request->request_pairs, attr_state);
+		fr_value_box_memdup(vp, &vp->data, NULL, eap_session->child_state,
+				    talloc_array_length(eap_session->child_state), false);
+		RDEBUG4("Restoring child state %pV", &vp->data);
 	}
 
-	return rcode;
-#endif
+	vp = fr_pair_find_by_da(&request->parent->request_pairs, NULL, attr_packet_type);
```

Comment:
Nit: this should always be Access-Request, right?  TTLS is pretty hard-coded to RADIUS.  :(